framework/wazuh/tests/test_agent.py

```diff
@@ -165,14 +165,19 @@ def test_agent_get_agents_summary_status(socket_mock, send_mock):
     summary = get_agents_summary_status(short_agent_list)
     assert isinstance(summary, WazuhResult), 'The returned object is not an "WazuhResult" instance.'
     # Asserts are based on what it should get from the fake database
-    expected_results = {'active': 2, 'disconnected': 1, 'never_connected': 1, 'pending': 1, 'total': 5}
+    expected_results = {'connection': {'active': 2, 'disconnected': 1, 'never_connected': 1, 'pending': 1, 'total': 5},
+                        'configuration': {'synced': 2, 'not_synced': 3, 'total': 5}}
     summary_data = summary['data']
-    assert set(summary_data.keys()) == set(expected_results.keys())
-    assert summary_data['active'] == expected_results['active']
-    assert summary_data['disconnected'] == expected_results['disconnected']
-    assert summary_data['never_connected'] == expected_results['never_connected']
-    assert summary_data['pending'] == expected_results['pending']
-    assert summary_data['total'] == expected_results['total']
+
+    # For the following test cases, if summary_data has unexpected keys, a KeyError will be raised
+
+    # Check the data dictionary follows the expected keys schema
+    assert all(summary_data[key].keys() == expected_results[key].keys() for key in summary_data.keys()), \
+        'The result obtained has unexpected keys'
+    # Check that the agents count for connection and configuration statuses are the expected ones
+    assert all(all(summary_data[key][status] == expected_results[key][status] for status in
+                   summary_data[key].keys()) for key in summary_data.keys()), \
+        'The agents connection or configuration status counts are not the expected ones'
```

Comment:
This kind of `assert` is problematic without a previous condition. For instance, if we add a test mutant in the tested function:

```python
@expose_resources(actions=["agent:read"], resources=["agent:id:{agent_list}"], post_proc_func=None)
def get_agents_summary_status(agent_list: list = None):
    """Count the number of agents by connection and groups configuration synchronization statuses.

    Parameters
    ----------
    agent_list : list[str]
       List of agents ID's

    Returns
    -------
    WazuhResult
    """
    return WazuhResult({'data': {}})
```

The test will keep passing when this is not correct:

```shellsession
============================= test session starts ==============================
collecting ... collected 1 item

test_agent.py::test_agent_get_agents_summary_status PASSED               [100%]

======================== 1 passed, 10 warnings in 0.93s ========================
```

Reply:
Changing the generator we are iterating will solve this issue.

Right now, I am iterating `summary_data.keys()`. It is here where, if `summary_data` is empty (case you commented (`{}`)), the test will wrongly pass.

If I use `expected_results.keys()` instead, the test will raise an `Exception` if `summary_data` doesn't have the same keys as `expected_results`. An `Exception` will also be raised if `summary_data` has more keys than `expected_results`.

Another solution could be adding `assert summary_data.keys() == expected_results.keys()` before these asserts.

Reply:
I always use the second solution with this kind of `assert`. Anyway, tests are best when you read them and know what they are doing in an instant. Using these structures with multiple `for` loops make that harder. Keep that in mind for further ones.